Zend/tests/type_declarations/static_type_return.phpt

```diff
@@ -0,0 +1,91 @@
+--TEST--
+Static return type
+--FILE--
+<?php
+
+class A {
+    public function test1(): static {
+        return new static;
+    }
+    public function test2(): static {
+        return new self;
+    }
+    public function test3(): static {
+        return new static;
+    }
+    public function test4(): static|array {
+        return new self;
+    }
+}
+
+class B extends A {
+    public function test3(): static {
+        return new C;
+    }
+}
+
+class C extends B {}
+
+$a = new A;
+$b = new B;
+
+var_dump($a->test1());
+var_dump($b->test1());
+
+echo "\n";
+var_dump($a->test2());
+try {
+    var_dump($b->test2());
+} catch (TypeError $e) {
+    echo $e->getMessage(), "\n";
+}
+
+echo "\n";
+var_dump($a->test3());
+var_dump($b->test3());
+
+echo "\n";
+var_dump($a->test4());
+try {
+    var_dump($b->test4());
+} catch (TypeError $e) {
+    echo $e->getMessage(), "\n";
+}
+
+echo "\n";
+$test = function($x): static {
+    return $x;
+};
+
+try {
+    var_dump($test(new stdClass));
+} catch (TypeError $e) {
+    echo $e->getMessage(), "\n";
+}
+
+$test = $test->bindTo($a);
+var_dump($test($a));
+
+?>
+--EXPECT--
+object(A)#3 (0) {
+}
+object(B)#3 (0) {
+}
+
+object(A)#3 (0) {
+}
+Return value of A::test2() must be an instance of static, instance of A returned
```

Comment:
This might be confusing at runtime in more complicated code - the error message gives no indication of what `static` is. (e.g. if a method returning static has incorrect caching based on $obj->getKey() which can be reused)

An error message mentioning what the current class scope is may save debugging time, e.g. `Return value of A::test2() must be an instance of static (currently static is %s(class/trait/interface) B), instance of A returned`

-----

Also, what should happen for code such as this? I assume it should be the same TypeError message, because `instanceof` only applies to classes and interfaces. (this is also a potential test case)

```
trait T {
    public static function f() : static {
        return new A();
    }
}
class A {
    use T;
}
T::f();
```

Reply:
I've changed the implementation to display the type `static` resolves to instead. We already do this for `self` and `parent`, so it makes sense to me to treat `static` the same way.